datafusion/physical-expr/src/expressions/dynamic_filters.rs

```diff
@@ -335,22 +313,12 @@ mod test {
         ]));
         // Each ParquetExec calls `with_new_children` on the DynamicFilterPhysicalExpr
         // and remaps the children to the file schema.
-        let dynamic_filter_1 = reassign_predicate_columns(
-            Arc::clone(&dynamic_filter) as Arc<dyn PhysicalExpr>,
-            &filter_schema_1,
-            false,
-        )
-        .unwrap();
-        let snap = dynamic_filter_1.snapshot().unwrap().unwrap();
-        insta::assert_snapshot!(format!("{snap:?}"), @r#"BinaryExpr { left: Column { name: "a", index: 0 }, op: Eq, right: Literal { value: Int32(42) }, fail_on_overflow: false }"#);
-        let dynamic_filter_2 = reassign_predicate_columns(
-            Arc::clone(&dynamic_filter) as Arc<dyn PhysicalExpr>,
-            &filter_schema_2,
-            false,
-        )
-        .unwrap();
-        let snap = dynamic_filter_2.snapshot().unwrap().unwrap();
-        insta::assert_snapshot!(format!("{snap:?}"), @r#"BinaryExpr { left: Column { name: "a", index: 1 }, op: Eq, right: Literal { value: Int32(42) }, fail_on_overflow: false }"#);
+        let dynamic_filter_1 = dynamic_filter.with_schema(Arc::clone(&filter_schema_1));
+        let snap_1 = dynamic_filter_1.snapshot().unwrap().unwrap();
+        insta::assert_snapshot!(format!("{snap_1:?}"), @r#"BinaryExpr { left: Column { name: "a", index: 0 }, op: Eq, right: Literal { value: Int32(42) }, fail_on_overflow: false }"#);
+        let dynamic_filter_2 = dynamic_filter.with_schema(Arc::clone(&filter_schema_2));
```

Comment:
What is expected to call `with_schema`? This seems like a new method on `DynamicFilterPhysicalExpr`. We need `reassign_predicate_columns` to work, that's what gets called from within `ParquetSource`, etc.

Reply:
When you need `reassign_predicate_columns`, basically it re-project columns based on the provided schema.

Reply:
Right but the constraint is that we can't modify what `reassign_predicate_columns` and similar do internally, otherwise that's a ton of API churn. The existing design works within the confines of the existing APIs.

Reply:
> We need reassign_predicate_columns to work, that's what gets called from within ParquetSource, etc

We can also change `reassign_predicate_columns` inside Parquet if that bring us to the better state. From my view, `reassign_predicate_columns` is the root cause why you ends up `with_new_children` that doesn't actually updating the "children".

Reply:
> otherwise that's a ton of API churn.

We only use in `DatafusionArrowPredicate`, is there any other places we need to change?

The main point is that `with_new_children` in the main branch isn't doing the right thing, it should update the source filter instead, but you only update the remapped filter schema. I think the filter schema is "parameters" for remapping column indexes, it doesn't need to be part of the filter expressions at all.

Reply:
If you really want to keep `reassign_predicate_columns` for whatever reason, you should pass the `inner` of `DynamicFilterPhysicalExpr` instead, so you are only modifying the `inner` and not the whole `DynamicFilterPhysicalExpr`. The difference is that the `with_new_children` is not called on the `DynamicFilterPhysicalExpr` level.